src/main/java/com/codessquad/qna/domain/Answer.java

```diff
@@ -0,0 +1,74 @@
+package com.codessquad.qna.domain;
+
+import org.hibernate.annotations.OnDelete;
+import org.hibernate.annotations.OnDeleteAction;
+
+import javax.persistence.*;
+import java.time.LocalDateTime;
+
+@Entity
+public class Answer {
+
+    @Id
+    @GeneratedValue
+    private Long id;
+
+    @ManyToOne
+    @OnDelete(action = OnDeleteAction.CASCADE)
```

Comment:
네, 코멘트에 이어서...
`Hibernate` 가 어떤 방식으로 `Cascade` 를 구현하는지 조금 더 조사해보실 것을 강력히 추천 드립니다.

Reply:
이건 제가 JPA와 Hibernate를 모르고 문제를 해결하기 위해 찾아보고 적용한 것입니다. 내부 구현이 어떻게 되어있는지는 잘 모릅니다. 그래서 일단은 JPA와 Hibernate에 대해서 간단히 공부하였습니다.

JPA는 Java EE에서 ORM 의 명세(Spec)입니다. 그리고 Hibernate는 JPA의 구현체 중 하나입니다.

JPA는 [Java Persistence API](https://www.oracle.com/java/technologies/persistence-jsp.html)라고 하여 POJO를 관계형 모델로 매핑하는 API이라고 합니다. 

영속성(Persistence)에 대해서는 아직까지 잘 모르겠습니다. 일단은 객체를 관계형 모델로 표현한 후에 DB에 지속적으로 저장되는 수준으로만 이해하고 있습니다.

Entity는 RDB의 테이블(관계형 모델의 Relation)으로 매핑이 되고, Entity와 Entity 사이에는 OneToOne, OneToMany, ManyToOne의 관계를 가질 수 있습니다. 이는 RDBMS의 외부키 참조로 구현되는 것 같습니다. 

Entity와 Entity 사이의 관계는 단방향일 수도, 양방향일 수도 있습니다. JPA에서 Entity 사이의 관계를 하나 선언하면, Entity에서 Field로의 방향으로 관계를 정의할 수 있습니다.

Cascade는 Entity와 Entity 사이에서 한쪽의 영속성이 변할 때 대상 방향으로의 영속성에도 변화가 전파되는 것을 의미합니다.

JPA에서는 OneToOne, OneToMany, ManyToMany의 cascade=CascadeType.--- 로 그 형태를 지정해 줄 수 있습니다.

JPA의 CascadeType은 다음과 같습니다.

- `PERSIST`
  - 부모가 영속성을 가지게 되면, 자식도 같이 영속성을 가짐
- `MERGE`
  - 부모와 자식이 값이 바뀐 후, 부모의 값이 Persistent Context에 저장되면, 자식의 변화도 Persistent Context에 반영됨.
- `REMOVE`
  - 부모가 DB에서 삭제되면, 자식도 DB에서 삭제됨
- `DETACH`
  - 부모가 Persistent Context에서 떨어지면, 자식도 Persistent Context에 떨어짐
- `REFRESH`
  - Refresh는 Persistent Context에서 값을 다시 가져오는 것인데, 부모만 가져와도 그 자식의 값도 다시 가져옴

DETACH와 REMOVE의 차이는 아직은 잘 모르겠는데, DB에서 삭제되는 것이 REMOVE면, DETACH는 JPA에서 관리하는 Persistent Context에서만 사라지는 것 정도의 차이로 이해했습니다.

Hibernate에 추가로 있는 Cascade Type이 있습니다. 이는 Hibernate의 Cascade 애노테이션으로 추가할 수 있으며, 다음과 같습니다.

- `LOCK`
  - DETACH를 다시 LOCK하는 것인데, 부모가 LOCK되면 자식도 LOCK
- `REPLICATE`
  - 어플리케이션에서 새로 만들어서 Persistent Context에 replicate 하면, 자식도 같이 실행됨.
- `SAVE_UPDATE`
  - DB에 부모가 Insert나 Update가 되면 자식도 Insert나 Update 됨.

아직 JPA의 각 동작이 정확하게 어떻게 되는지를 잘 몰라서 추가적으로 공부를 해야할 것 같습니다.

추가로 Hibernate의 [OnDelete](https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#pc-cascade-on-delete) 애노테이션이 있습니다. 이것은 DDL 레벨에서 
```sql
alter table Phone
    add constraint FK82m836qc1ss2niru7eogfndhl
    foreign key (owner_id)
    references Person
    on delete cascade
```
와 같은 식의 추가적으로 외부키 제약을 추가해준다고 합니다.

>Without the @OnDelete annotation, the @OneToMany association relies on the cascade attribute to propagate the remove entity state transition from the parent entity to its children. However, when the @OnDelete annotation is in place, Hibernate prevents the child entity DELETE statement from being executed while flushing the Persistence Context.
>This way, only the parent entity gets deleted, and all the associated child records are removed by the database engine, instead of being deleted explicitly via DELETE statements.

라고 해서, OnDelete는 Persistence Context를 flush할 때 자식 Entity를 삭제하는 쿼리를 보내지 않고 DB 엔진에 맡긴다고 되어있습니다. 실제로 Question를 삭제하면 삭제 전에 Answer를 삭제하는 쿼리가 생기지 않습니다.

그래서 OnDelete를 적용하지 않고, Question에 `OneToMany(cascade = CascadeType.REMOVE)`로 Answer와 관계를 맺고, Question을 삭제했는데, Question을 삭제하기 전에 Answer를 하나씩 anwserId로 삭제하는 것을 확인했습니다.

또한, 단순히 여러개의 Answer를 Spring Data JPA로 Question으로 삭제해보기도 했습니다. 그 결과 questionId로 Answer를 삭제하는 쿼리가 하나 날아가는 것이 아니라, answerId로 하나씩 여러개를 삭제하는 쿼리가 생겼습니다.

제가 Hibernate의 동작 코드를 본 것도 아니고, 공부한 것도 아닙니다. 그래서 추측하건데, Persistent Context와 DB가 완전하게 동일하지는 않고, Persistent Context에서 Entity들을 관리하는데, 그 내부에서 List같은 것으로 관리를 하기 때문에 하나씩 탐색하며 삭제를 할 것 같습니다. 그래서 OnDelete로 DB에 제약관계를 주더라도, DB로 보내는 쿼리만 발생하지 않을 뿐, Persistent Context를 DB와 동일한 상태로 유지하기 위해서 하나씩 탐색하면서 삭제를 할 것 같습니다.

아직 지식이 부족해서, 추가로 JPA에 대해서 공부해야할 것 같습니다.